packages/terra-date-picker/src/DateInput.jsx

```diff
@@ -99,124 +107,330 @@ const defaultProps = {
+   * corrections easier, and is re-enabled if the whole date is erased.
+   * @param {string} inputValue - The value from the current input.
+   * @param {int} type - The input type, based on DateUtil.inputType.
+   * @param {string} dateVariant - Date variant based on the currently supported date
+   *                      orders: 'MM-DD-YYYY', 'DD-MM-YYYY', or 'YYYY-MM-DD'.
+   */
+  const moveFocusOnChange = (inputValue, type, dateVariant) => {
+    if (dateVariant === 'MM-DD-YYYY') {
+      if (inputValue.length === 2) {
+        if (type === DateUtil.inputType.MONTH) {
+          dayFocus.focus();
+        } else {
+          yearFocus.focus();
+        }
+      }
+    } else if (dateVariant === 'DD-MM-YYYY') {
+      if (inputValue.length === 2) {
+        if (type === DateUtil.inputType.DAY) {
+          monthFocus.focus();
+        } else {
+          yearFocus.focus();
+        }
+      }
+    } else if (dateVariant === 'YYYY-MM-DD') {
+      if (inputValue.length === 2 && type === DateUtil.inputType.MONTH) {
+        dayFocus.focus();
+      } else if (inputValue.length === 4) {
+        monthFocus.focus();
+      }
     }
-  }
+  };
 
-  handleOnChange(event) {
-    if (!DateUtil.validDateInput(event.target.value)) {
-      return;
+  /**
+   * Sets state for the day, month, and year.
+   * @param {string} inputValue - The input value to set in state.
+   * @param {int} type - The inputType (day, month, or year).
+   */
+  const setDate = (inputValue, type) => {
+    if (type === DateUtil.inputType.DAY) {
+      dateDispatch({ day: inputValue });
+    } else if (type === DateUtil.inputType.MONTH) {
+      dateDispatch({ month: inputValue });
+    } else {
+      dateDispatch({ year: inputValue });
     }
+    moveFocusOnChange(inputValue, type, localeFormat);
+  };
+
+  const handleDateChange = (event, type) => {
+    const inputValue = event.target.value.replace(/\D/gm, '');
 
-    if (this.props.onChange) {
-      this.props.onChange(event);
+    /**
+     * Sets the day, month and year based on input values, formats them
+     * based on the date format variant, and passes the formatted date to onChange.
+     */
+    const day = type === DateUtil.inputType.DAY ? inputValue : date.day;
+    const month = type === DateUtil.inputType.MONTH ? inputValue : date.month;
+    const year = type === DateUtil.inputType.YEAR ? inputValue : date.year;
+    let inputDate;
+    let formattedDate;
+    if (day.length === 2 && month.length === 2 && year.length === 4) {
+      inputDate = DateUtil.formatISODate(`${year}-${month}-${day}`, 'YYYY-MM-DD');
+      formattedDate = DateUtil.strictFormatISODate(inputDate, dateFormat);
     }
-  }
 
-  handleOnKeyDown(event) {
-    if (this.props.onKeyDown) {
-      this.props.onKeyDown(event);
+    /* eslint-disable no-param-reassign */
+    if (DateUtil.isValidDate(formattedDate, dateFormat)) {
+      event.target.value = formattedDate;
```

Comment:
Why are we setting the formatted date back on the input?

Reply:
The actual value is only the most recently entered part of the date. The logic in the `DatePicker.jsx` is looking for a single input that eventually turns into a whole date (and really, react-datepicker itself is also built entirely around a single input). This way, I can leave the `DatePicker.jsx` (and react-datepicker itself) as-is without major changes, still allow raw input, and also pass it a full date when it expects it. This was (mostly) done to try to contain the various changes needed into a single place as possible.